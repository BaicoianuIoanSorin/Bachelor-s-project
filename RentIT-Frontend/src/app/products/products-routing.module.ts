import {NgModule} from '@angular/core';
import {RouterModule, Routes} from '@angular/router';
import {PRODUCTS_MENU_ITEM_URLS} from "src/app/constants";
import {AddingProductsComponent} from "src/app/products/adding-products/adding-products.component";
import {ProductsModulePage} from "src/app/products/products.module.page";
import {ProductsComponent} from "src/app/products/products/products.component";
import {ProductComponent} from "src/app/products/product/product/product.component";

const routes: Routes = [
    {
      path: '',
      component: ProductsModulePage,
      children: [
        {
          title: 'Products',
          path: '',
          component: ProductsComponent,
        },
        {
          title: 'Add product',
          path: PRODUCTS_MENU_ITEM_URLS.ADDING_PRODUCTS.substring(1),
          component: AddingProductsComponent
        },
        {
          title: 'Product overview',
          path: PRODUCTS_MENU_ITEM_URLS.PRODUCT.substring(1),
          loadChildren: () =>
            import('./product/product.module').then((m) => m.ProductModule),
        }
      ]
    },
  ]
;

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class ProductsRoutingModule {
}
