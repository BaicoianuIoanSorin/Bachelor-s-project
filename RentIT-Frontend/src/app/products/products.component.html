<nb-card
  [nbSpinner]="isFetching$ | async" nbSpinnerSize="large" nbSpinnerStatus="primary">
  <nb-card-header>
    <h1> Here are some products for you</h1>
  </nb-card-header>
  <nb-card-body *ngIf="products$ | async as products">
    <product *ngFor="let product of products"
             [product]="product"
    ></product>
  </nb-card-body>
  <nb-card-body>
  </nb-card-body>
</nb-card>
