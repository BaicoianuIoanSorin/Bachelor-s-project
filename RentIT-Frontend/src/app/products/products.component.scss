.product-grid {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  padding: 0 30px;
}

/* Apply margin to the last item in each row */
.last-in-row {
  margin-right: 0;
}
