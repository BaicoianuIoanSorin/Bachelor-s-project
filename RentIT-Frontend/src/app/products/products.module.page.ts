import {Component, OnDestroy, OnInit} from "@angular/core";

@Component({
  template: `<router-outlet></router-outlet>`,
})
export class ProductsModulePage implements OnInit, OnDestroy {
  alive = true;

  constructor() {
  }

  ngOnInit() {

  }

  ngOnDestroy() {
    this.alive = false;
  }
}
