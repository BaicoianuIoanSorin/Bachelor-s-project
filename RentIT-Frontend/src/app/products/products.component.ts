import {Component, OnDestroy, OnInit} from '@angular/core';
import {Select, Store} from "@ngxs/store";
import {ProductsFetch, ProductsReset} from "src/app/products/products.actions";
import {ProductsSelector} from "src/app/products/products.selector";
import {Observable} from "rxjs";
import {Product} from "src/model/product";

@Component({
  selector: 'app-products',
  templateUrl: './products.component.html',
  styleUrls: ['./products.component.scss']
})
export class ProductsComponent implements OnInit, OnDestroy {
  @Select(ProductsSelector.isFetching)
  isFetching$: Observable<boolean>;
  @Select(ProductsSelector.products)
  products$: Observable<Product[]>

  alive : boolean = true;

  constructor(
    private store: Store,
  ) {
  }

  ngOnInit(): void {
    this.store.dispatch(new ProductsFetch());
  }

  ngOnDestroy(): void {
    this.alive = false;
    this.store.dispatch(new ProductsReset());
  }
}
