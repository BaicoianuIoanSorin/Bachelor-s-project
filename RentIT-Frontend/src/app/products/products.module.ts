import { NgModule } from '@angular/core';
import {CommonModule, NgOptimizedImage} from '@angular/common';
import { ProductsRoutingModule } from './products-routing.module';
import { NgxsModule } from '@ngxs/store';
import { ProductsState } from './products.state';
import { ProductsComponent } from './products.component';
import { ProductComponent } from './components/product/product.component';
import { NbButtonModule, NbCardModule, NbIconModule, NbSpinnerModule, NbTooltipModule } from '@nebular/theme';
import { NbEvaIconsModule } from '@nebular/eva-icons';
import {FormsModule} from "@angular/forms";

export const PRODUCTS_NG_MODULES = [
  NgOptimizedImage,
]

export const PRODUCTS_NEBULAR_COMPONENTS = [
  NbButtonModule,
  NbEvaIconsModule,
  NbCardModule,
  NbIconModule,
  NbSpinnerModule,
  NbTooltipModule,
];

export const PRODUCTS_STATES = [
  ProductsState,
];

export const PRODUCTS_COMPONENTS = [
  ProductsComponent,
  ProductComponent,
];

@NgModule({
  declarations: [
    PRODUCTS_COMPONENTS,
  ],
  imports: [
    CommonModule,
    ProductsRoutingModule,
    NgxsModule.forFeature(PRODUCTS_STATES),
    ...PRODUCTS_NEBULAR_COMPONENTS,
    ...PRODUCTS_NG_MODULES,
  ]
})
export class ProductsModule { }
