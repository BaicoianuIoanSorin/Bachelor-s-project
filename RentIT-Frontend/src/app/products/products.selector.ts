import {Selector} from "@ngxs/store";
import {ProductsState, ProductsStateModel} from "src/app/products/products.state";

export class ProductsSelector {
  @Selector([ProductsState])
  static isFetching(state: ProductsStateModel) {
    return state.isFetching;
  }

  @Selector([ProductsState])
  static products(state: ProductsStateModel) {
    return state.products;
  }

  @Selector([ProductsState])
  static endOfList(state: ProductsStateModel) {
    return state.endOfList;
  }
}
