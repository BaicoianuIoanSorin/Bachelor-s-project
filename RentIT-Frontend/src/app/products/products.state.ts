import {Product} from "src/model/product";
import {Action, State, StateContext} from "@ngxs/store";
import {Injectable} from "@angular/core";
import {NbToastrService} from "@nebular/theme";
import {ProductsFetch, ProductsReset} from "src/app/products/products.actions";
import {produce} from "immer";
import {environment} from "src/environment/environment.dev";
import {ICONS} from "src/app/constants";
import {mockedProducts} from "src/mocks/products.mock";

export interface ProductsStateModel {
  isFetching: boolean;
  products: Product[];
}

export const defaultsState: ProductsStateModel = {
  isFetching: false,
  products: [],
}

@State<ProductsStateModel>({
  name: 'productsPage',
  defaults: defaultsState,
})

@Injectable()
export class ProductsState {
  constructor(
    private toastrService: NbToastrService,
  ) {
  }

  @Action(ProductsFetch)
  async moviesFetchNextPage(
    {getState, setState}: StateContext<ProductsStateModel>) {

    let newState = produce(getState(), draft => {
      draft.isFetching = true;
    })
    setState(newState);

    // TODO here the service api will be called and fetch the actual products, for now mock

    try {
      // mock
      const products = mockedProducts;

      newState = produce(getState(), draft => {
        draft.products = products;
      })
      setState(newState);
    } catch (error) {
      this.toastrService.danger(
        environment.production ? 'Please contact the administration' : error,
        'Something went wrong',
        {icon: ICONS.ALERT_CIRCLE_OUTLINE}
      );
      console.error('error');
    }
  }

  @Action(ProductsReset)
  async productsReset(
    { setState }: StateContext<ProductsStateModel>) {
   setState(defaultsState);
  };
}
