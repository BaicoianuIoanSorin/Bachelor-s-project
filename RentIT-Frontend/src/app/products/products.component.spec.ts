import { Spectator } from "@ngneat/spectator";
import { createComponentFactory } from "@ngneat/spectator/jest";
import { NgxsModule, Store } from "@ngxs/store";
import { RouterTestingModule } from "@angular/router/testing";
import { BrowserModule } from "@angular/platform-browser";
import { BrowserAnimationsModule } from "@angular/platform-browser/animations";
import { HttpClientModule } from "@angular/common/http";
import { of } from "rxjs";
import {ProductsComponent} from "./products.component";
import {AppRoutingModule} from "../app-routing.module";
import {NEBULAR_MODULES, PROVIDERS, STATES} from "../app.module";

// const productsServiceMock = {
//   getProducts: jest.fn().mockReturnValue(of([])),
// };

describe('app > products > products.component.spec.ts', () => {
  let spectator: Spectator<ProductsComponent>;

  const createComponent = createComponentFactory({
    component: ProductsComponent,
    imports: [
      RouterTestingModule,
      BrowserModule,
      HttpClientModule,
      AppRoutingModule,
      BrowserAnimationsModule,
      NgxsModule.forRoot(STATES),
      ...NEBULAR_MODULES,
    ],
    providers: [
      ...PROVIDERS,
      Store,
      // {
      //   provide: ProductsService,
      //   useValue: productsServiceMock
      // },
    ],
  });

  beforeEach(() => {
    spectator = createComponent();
  });

  it('should create', () => {
    expect(spectator.component).toBeTruthy();
  });

  it('should match snapshot', async () => {
    spectator.detectChanges();
    await spectator.fixture.whenStable();
    spectator.detectChanges();
    await spectator.fixture.whenRenderingDone();

    // assert
    expect(spectator.fixture).toMatchSnapshot();
  });

  // TODO add test later when data is available
});
