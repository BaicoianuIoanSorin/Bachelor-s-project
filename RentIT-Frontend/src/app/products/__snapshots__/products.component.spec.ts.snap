// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app > products > products.component.spec.ts should match snapshot 1`] = `
<app-main-page>
  <p>
    main-page works!
  </p>
</app-main-page>
`;
